internal/config/variables/variables.go

```diff
@@ -329,23 +368,168 @@ func LoadEnvValues(vars map[string]*Variable) ([]*pb.Variable, hcl.Diagnostics)
+) bool {
+	// Get all our variables with dynamic defaults
+	dynamicVars := map[string]*Variable{}
+	for k, v := range vars {
+		if v.Default != nil {
+			val := v.Default.Value
+			if val.Type() == dynamic.Type {
+				dynamicVars[k] = v
+			}
+		}
+	}
+
+	// Go through our variable values and delete any dynamic vars we have
+	// values for already; we do not need to fetch those.
+	for _, pbv := range pbvars {
+		delete(dynamicVars, pbv.Name)
+	}
+
+	return len(dynamicVars) > 0
+}
+
+// LoadDynamicDefaults will load the default values for variables that have
+// dynamic configurations. This will only load the values if there isn't an
+// existing variable set in pbvars. Therefore, it is recommended that this is
+// called last and the values are _prepended_ to pbvars for priority.
+func LoadDynamicDefaults(
+	ctx context.Context,
+	log hclog.Logger,
+	pbvars []*pb.Variable,
+	vars map[string]*Variable,
+	dynamicOpts ...appconfig.Option,
+) ([]*pb.Variable, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+
+	// Get all our variables with dynamic defaults
+	dynamicVars := map[string]*Variable{}
+	for k, v := range vars {
+		if v.Default != nil {
+			val := v.Default.Value
+			if val.Type() == dynamic.Type {
+				dynamicVars[k] = v
+			}
+		}
+	}
+
+	// If we have no dynamic vars, we do nothing.
+	if len(dynamicVars) == 0 {
+		return nil, diags
+	}
+
+	// Go through our variable values and delete any dynamic vars we have
+	// values for already; we do not need to fetch those.
+	for _, pbv := range pbvars {
+		delete(dynamicVars, pbv.Name)
+	}
+
+	// If we have no dynamic vars we need values for, also do nothing.
+	if len(dynamicVars) == 0 {
+		return nil, diags
+	}
+
+	// Build our watcher
+	ch := make(chan *appconfig.UpdatedConfig)
+	w, err := appconfig.NewWatcher(append(
+		append([]appconfig.Option{}, dynamicOpts...),
+		appconfig.WithLogger(log),
+		appconfig.WithNotify(ch),
+	)...)
+	if err != nil {
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Error initializing configuration watcher",
+			Detail:   err.Error(),
+			Subject: &hcl.Range{
+				Filename: "waypoint.hcl",
+			},
+		})
+		return nil, diags
+	}
+	defer w.Close()
+
+	// We have some variables we need to fetch dynamic values for.
+	configVars := make([]*pb.ConfigVar, 0, len(dynamicVars))
+	for k, v := range dynamicVars {
+		configVars = append(configVars, &pb.ConfigVar{
+			Name: k,
+			Value: &pb.ConfigVar_Dynamic{
+				Dynamic: v.Default.Value.EncapsulatedValue().(*pb.ConfigVar_DynamicVal),
+			},
+
+			// This is set to true on purpose because it forces the appconfig
+			// watcher to give us an easier to consume format (struct vs
+			// array of key=value strings for env vars).
+			NameIsPath: true,
+		})
+	}
+
+	// Send our variables. Purposely ignore the error return value because
+	// it can only ever be a context cancellation which we pick up in the
+	// select later.
+	w.UpdateVars(ctx, configVars)
+
+	// Wait for values.
+	select {
```

Comment:
Can we add a timeout to this loop that will emit some log or UI change that will notify the user of the delay? I'm thinking about instances where, say, a vault server isn't available for some reason and this select hangs for an extended period of time without the user knowing why.

Reply:
There's already considerable logging around the fetching of values, so its not going to look like its doing nothing. Adding an additional log here though after a delay makes sense. For timing out, I think upstream should just set a context timeout.